### PR TITLE
Added multimessage serializer

### PR DESCRIPTION
**Pull request by dartartem**

### CODE REVIEW

File: eventuate-messaging-kafka-basic-consumer/src/test/resources/application.properties [hunk 1]
```diff
@@ -2,7 +2,7 @@ eventuate.local.kafka.consumer.properties.session.timeout.ms=30000
 eventuate.local.kafka.consumer.properties.key.serializer=org.apache.kafka.common.serialization.StringSerializer
 
 eventuate.local.kafka.producer.properties.buffer.memory=1000000
-eventuate.local.kafka.producer.properties.value.serializer=org.apache.kafka.common.serialization.StringSerializer
+eventuate.local.kafka.producer.properties.value.serializer=org.apache.kafka.common.serialization.ByteArraySerializer
```
Comment (eventuateio) [inline eventuate-messaging-kafka-basic-consumer/src/test/resources/application.properties]: Are you sure that this should this be configurable?

File: eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageBuilder.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      binaryStream.write(MAGIC_ID_BYTES);
+      size += MAGIC_ID_BYTES.length;
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public boolean addMessage(EventuateKafkaMultiMessageKeyValue message) {
+    try {
+      byte[] keyBytes = Optional.ofNullable(message.getKey()).map(String::getBytes).orElse(new byte[0]);
```
Comment (eventuateio) [inline eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageBuilder.java]: getBytes() should be passed a CharSet to ensure cross-platform consistency

File: eventuate-messaging-kafka-basic-consumer/src/test/resources/application.properties (hunk 1 above)
Comment (dartartem) [reply eventuate-messaging-kafka-basic-consumer/src/test/resources/application.properties]: It is just a test for property parsing. We can add different kafka configs, including this.
To make it not configurable, we need to add direct blocking (checks with exceptions) for such properties.

File: eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageBuilder.java (hunk 2 above)
Comment (dartartem) [reply eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageBuilder.java]: yes, I also thought about, so I will fix it tomorrow everywhere.

File: eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  public static final byte[] MAGIC_ID_BYTES = MAGIC_ID.getBytes();
+
+  public byte[] convertMessagesToBytes(List<EventuateKafkaMultiMessageKeyValue> messages) {
+    ByteArrayOutputStream binaryStream = new ByteArrayOutputStream();
+
+    try {
+      binaryStream.write(MAGIC_ID_BYTES);
+
+      for (EventuateKafkaMultiMessageKeyValue message : messages) {
+        byte[] keyBytes = Optional.ofNullable(message.getKey()).map(String::getBytes).orElse(new byte[0]);
```
Comment (eventuateio) [inline eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java]: Shouldn't this use EventuateKafkaMultiMessageBuilder?

File: eventuate-messaging-kafka-consumer/src/main/java/io/eventuate/messaging/kafka/consumer/MessageConsumerKafkaImpl.java [hunk 4]
```diff
@@ -73,4 +83,10 @@ public void close() {
   public String getId() {
     return id;
   }
+
+  private boolean isMultiMessage(RawKafkaMessage message) {
+    return message.getPayload().length >= EventuateKafkaMultiMessageConverter.MAGIC_ID_BYTES.length
+            && Arrays.equals(EventuateKafkaMultiMessageConverter.MAGIC_ID_BYTES,
+                    Arrays.copyOf(message.getPayload(), EventuateKafkaMultiMessageConverter.MAGIC_ID_BYTES.length));
```
Comment (eventuateio) [inline eventuate-messaging-kafka-consumer/src/main/java/io/eventuate/messaging/kafka/consumer/MessageConsumerKafkaImpl.java]: This should not copy the array

File: eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java (hunk 3 above)
Comment (dartartem) [reply eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java]: Probably this method is not needed. As I wrote in email, there is some copy paste. I will need to make a clean up.

File: eventuate-messaging-kafka-consumer/src/main/java/io/eventuate/messaging/kafka/consumer/MessageConsumerKafkaImpl.java (hunk 4 above)
Comment (dartartem) [reply eventuate-messaging-kafka-consumer/src/main/java/io/eventuate/messaging/kafka/consumer/MessageConsumerKafkaImpl.java]: Ok, I will compare it in different way

File: eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageBuilder.java (hunk 2 above)
Comment (eventuateio) [reply eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageBuilder.java]: Similarly, new String(..) or equivalent needs to use the same Charset

Comment (dartartem) [reply eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageBuilder.java]: ok

File: eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    }
+    else {
+      return Collections.singletonList(new String(bytes, Charset.forName("UTF-8")));
+    }
+  }
+
+  public boolean isMultiMessage(byte[] message) {
+    if (message.length < MAGIC_ID_BYTES.length) return false;
+
+    return ByteBuffer.wrap(message, 0, MAGIC_ID_BYTES.length).equals(ByteBuffer.wrap(MAGIC_ID_BYTES));
```
Comment (cer) [inline eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java]: This is just a for loop.

File: eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+    for (EventuateKafkaMultiMessageKeyValue message : messages) {
+      builder.addMessage(message);
+    }
+
+    return builder.toBinaryArray();
+  }
+
+  public List<EventuateKafkaMultiMessageKeyValue> convertBytesToMessages(byte[] bytes) {
+    ByteArrayInputStream binaryStream = new ByteArrayInputStream(bytes);
```
Comment (cer) [inline eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java]: Use ByteByffer.wrap(..) instead?

Comment (dartartem) [reply eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java]: ok

File: eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java (hunk 5 above)
Comment (dartartem) [reply eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java]: ok, I will use it directly

**Pull request merged**